lib/backend/pgbk/common/gcp.go

```diff
@@ -0,0 +1,139 @@
+	return gcpOAuthAccessTokenBeforeConnect(ctx, gcpAccessTokenGetterImpl{}, gcpSQLOAuthScope, logger)
+}
+
+// GCPAlloyDBBeforeConnect returns a pgx BeforeConnect function suitable for GCP
+// AlloyDB (PostgreSQL-compatible) with IAM authentication.
+func GCPAlloyDBBeforeConnect(ctx context.Context, logger *slog.Logger) (func(ctx context.Context, config *pgx.ConnConfig) error, error) {
+	return gcpOAuthAccessTokenBeforeConnect(ctx, gcpAccessTokenGetterImpl{}, gcpAlloyDBOAuthScope, logger)
+}
+
+const (
+	// gcpSQLOAuthScope is the scope used for GCP SQL IAM authentication.
+	// https://developers.google.com/identity/protocols/oauth2/scopes#sqladmin
+	gcpSQLOAuthScope = "https://www.googleapis.com/auth/sqlservice.admin"
+	// gcpAlloyDBOAuthScope is the scope used for GCP AlloyDB IAM authentication.
+	// https://cloud.google.com/alloydb/docs/connect-iam
+	gcpAlloyDBOAuthScope = "https://www.googleapis.com/auth/alloydb.login"
+
+	gcpServiceAccountEmailSuffix = ".gserviceaccount.com"
+)
+
+type gcpAccessTokenGetter interface {
+	getFromCredentials(ctx context.Context, credentials *google.Credentials) (*oauth2.Token, error)
+	generateForServiceAccount(ctx context.Context, serviceAccount, scope string) (string, time.Time, error)
+}
+
+func gcpOAuthAccessTokenBeforeConnect(ctx context.Context, tokenGetter gcpAccessTokenGetter, scope string, logger *slog.Logger) (func(context.Context, *pgx.ConnConfig) error, error) {
+	defaultCred, err := google.FindDefaultCredentials(ctx, scope)
+	if err != nil {
+		// google.FindDefaultCredentials gives pretty error descriptions already.
+		return nil, trace.Wrap(err)
+	}
+
+	// This function tries to capture service account emails from various
+	// credentials methods but may fail for some unknown scenarios.
+	defaultServiceAccount, err := gcp.GetServiceAccountFromCredentials(defaultCred)
+	if err != nil || defaultServiceAccount == "" {
+		logger.WarnContext(ctx, "Failed to get service account email from default google credentials. Teleport will assume the database user in the PostgreSQL connection string matches the service account of the default google credentials.", "err", err, "sa", defaultServiceAccount)
+	}
+
+	return func(ctx context.Context, config *pgx.ConnConfig) error {
+		// IAM auth users have the PostgreSQL username of their emails minus the
+		// ".gserviceaccount.com" part. Now add the suffix back for the full
+		// service account email.
+		serviceAccountToAuth := config.User + gcpServiceAccountEmailSuffix
+
+		// If the requested db user is for another service account, the
+		// "host"/default service account can impersonate the target service
+		// account as a Token Creator. This is useful when using a different
+		// database user for change feed.
+		if defaultServiceAccount != "" && defaultServiceAccount != serviceAccountToAuth {
+			token, expires, err := tokenGetter.generateForServiceAccount(ctx, serviceAccountToAuth, scope)
```

Comment:
Does the `tokenGetter` do its own caching and singleflighting for generated tokens? Right after a restart or some other Big Cluster Event™ we're almost guaranteed to burst to the max conns of the pool, which may be like 50 IAM calls at once.

Reply:
`credentials.TokenSource.Token()` will only refresh token when necessary. 

For the case to `GenerateAccessToken`, i can add a FnCache.